collector/mesos_agent/mesos_agent.go

```diff
@@ -115,36 +149,109 @@ type resourceStatistics struct {
+	maColLog.Debugf("Fetching state from DC/OS host %s", host)
+	if err := h.getAgentState(); err != nil {
+		maColLog.Errorf("Failed to get agent state from %s. Error: %s", host, err)
+	}
+
+	maColLog.Debugf("Fetching container metrics from DC/OS host %s", host)
+	if err := h.getContainerMetrics(); err != nil {
+		maColLog.Errorf("Failed to get container metrics from %s. Error: %s", host, err)
+	}
+}
+
 // getContainerMetrics queries an agent for container-level metrics, such as
 // CPU, memory, disk, and network usage.
-func (h *DCOSHost) getContainerMetrics() ([]agentContainer, error) {
-	var containers []agentContainer
+func (h *MesosAgentCollector) getContainerMetrics() error {
+	h.containerMetrics = []agentContainer{}
 	u := url.URL{
-		Scheme: "http",
-		Host:   net.JoinHostPort(h.IPAddress, strconv.Itoa(h.Port)),
+		Scheme: h.RequestProtocol,
+		Host:   net.JoinHostPort(h.NodeInfo.IPAddress, strconv.Itoa(h.Port)),
 		Path:   "/containers",
 	}
 
 	h.HTTPClient.Timeout = HTTPTIMEOUT
 
-	return containers, Fetch(h.HTTPClient, u, &containers)
+	return http_client.Fetch(h.HTTPClient, u, &h.containerMetrics)
 }
 
 // getAgentState fetches the state JSON from the Mesos agent, which contains
 // info such as framework names and IDs, the current leader, config flags,
 // container (executor) labels, and more.
-func (h *DCOSHost) getAgentState() (agentState, error) {
-	state := agentState{}
+func (h *MesosAgentCollector) getAgentState() error {
+	h.agentState = agentState{}
 
 	u := url.URL{
-		Scheme: "http",
-		Host:   net.JoinHostPort(h.IPAddress, strconv.Itoa(h.Port)),
+		Scheme: h.RequestProtocol,
+		Host:   net.JoinHostPort(h.NodeInfo.IPAddress, strconv.Itoa(h.Port)),
 		Path:   "/state",
 	}
 
 	h.HTTPClient.Timeout = HTTPTIMEOUT
 
-	return state, Fetch(h.HTTPClient, u, &state)
+	return http_client.Fetch(h.HTTPClient, u, &h.agentState)
+}
+
+func (h *MesosAgentCollector) transformContainerMetrics() (out []producers.MetricsMessage) {
+	var msg producers.MetricsMessage
+	t := time.Unix(h.timestamp, 0)
+
+	// Produce container metrics
+	for _, c := range h.containerMetrics {
+		msg = producers.MetricsMessage{
+			Name:       producers.ContainerMetricPrefix,
+			Datapoints: buildDatapoints(c, t),
+			Timestamp:  t.UTC().Unix(),
+		}
+
+		fi, ok := getFrameworkInfoByFrameworkID(c.FrameworkID, h.agentState.Frameworks)
+		if !ok {
+			maColLog.Warnf("Did not find FrameworkInfo for framework ID %s, skipping!", fi.ID)
+			continue
+		}
+
+		msg.Dimensions = producers.Dimensions{
+			MesosID:            h.NodeInfo.MesosID,
+			ClusterID:          h.NodeInfo.ClusterID,
+			Hostname:           h.NodeInfo.Hostname,
+			ContainerID:        c.ContainerID,
+			ExecutorID:         c.ExecutorID,
+			FrameworkID:        c.FrameworkID,
+			FrameworkName:      fi.Name,
+			FrameworkRole:      fi.Role,
+			FrameworkPrincipal: fi.Principal,
+			Labels:             getLabelsByContainerID(c.ContainerID, h.agentState.Frameworks),
```

Comment:
`getLabelsByContainerID` should probably return error and you need to handle the error before you initiate `msg.Dimensions`